### ISSUE TITLE
Stock Fail to Update for Added/Removed Products in On-Hold Orders

### ISSUE THREAD

**Issue by erricgunawan** [woocommerce/woocommerce-ios#11800]
**Describe the bug**
When modifying an On-Hold order in the application by either adding or removing products, the stock levels do not automatically update to reflect these changes. The products are correctly added or removed from the order, but the corresponding adjustment in stock quantity is not triggered immediately.

***Workaround***
To trigger the stock adjustment, it is currently necessary to change the order status from 'On-Hold' to 'Pending Payment' and then revert it back to 'On-Hold'. 

Also, updating the On-Hold Order from WP-Admin correctly adjusts to the correct stock.

**To Reproduce**
1. Go to Orders and select an existing On-Hold Order (or create a new one)
2. Edit the order by either adding a new product or removing an existing one. Or adjust the number of items purchased for an existing product in the Order.
3. Observe that the product list in the order updates correctly, but the stock quantity does not adjust accordingly in the Order Notes.

**Screenshots**

<details>
  <summary>Screen Recording:</summary>
  <br>
  There's a video recording from the user in 7570597-zen but it's too big to upload here (241 MB).

  The video is also uploaded to p1706168218694079/1706168063.450169-slack-C6H8C3G23
</details>

**Expected behavior**
The stock quantity should automatically update to reflect the changes (addition/removal of products) in the on-hold order without requiring a change in order status or manual update intervention.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**
Please include:
- Device: iPhone 14 Pro Max
- iOS version: 17.3
- WooCommerce iOS version: 17.0

**WordPress Environment**
<details>

```
Not sure if it's still relevant, but if one is required, please ping me and I can provide one from my testing site.
```
</details>

**Comment by rachelmcr** [comment]
I took a quick look at this and wanted to share my findings:

* Stock levels are reduced in [wc_maybe_reduce_stock_levels\( $order_id \)](https://github.com/woocommerce/woocommerce/blob/b50431d9afda2660a78a64a03e6e10b0c79ef67f/plugins/woocommerce/includes/wc-stock-functions.php#L92), and that is triggered by the [woocommerce_order_status_on-hold](https://github.com/woocommerce/woocommerce/blob/b50431d9afda2660a78a64a03e6e10b0c79ef67f/plugins/woocommerce/includes/wc-stock-functions.php#L115C14-L115C46) hook for on-hold orders.
* That hook is called in [status_transition\(\)](https://github.com/woocommerce/woocommerce/blob/b50431d9afda2660a78a64a03e6e10b0c79ef67f/plugins/woocommerce/includes/class-wc-order.php#L411), i.e. when an order transitions to that status.
* This fits with the behavior we’re seeing in the app, where the stock levels are reduced when the status is changed. (`status_transition()` is used in [set_status()](https://github.com/woocommerce/woocommerce/blob/f46ffe14506ad63fec0be39d224c771b54993159/plugins/woocommerce/includes/class-wc-order.php#L285) which is called [when the order is updated via the API](https://github.com/woocommerce/woocommerce/blob/f46ffe14506ad63fec0be39d224c771b54993159/plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-orders-controller.php#L222).)

There must be a different core method that updates the stock levels when a product is added to an order on the web (and that isn't being triggered when the REST API updates the order), but I don't see where that is.

**Comment by rachelmcr** [comment]
I found the PR that added this behavior to core (https://github.com/woocommerce/woocommerce/pull/22329). This led me to how the stock is updated on the web in this scenario:

There's a function to [adjust the stock levels for line items](https://github.com/woocommerce/woocommerce/blob/f46ffe14506ad63fec0be39d224c771b54993159/plugins/woocommerce/includes/admin/wc-admin-functions.php#L227) added to an order, called when [line items are saved in the order](https://github.com/woocommerce/woocommerce/blob/f46ffe14506ad63fec0be39d224c771b54993159/plugins/woocommerce/includes/admin/wc-admin-functions.php#L302C10-L302C29). But that's only used when [order items are added via AJAX](https://github.com/woocommerce/woocommerce/blob/f46ffe14506ad63fec0be39d224c771b54993159/plugins/woocommerce/includes/class-wc-ajax.php#L1020) (on the web) so we can't trigger it from the app via the REST API.

As far as I'm seeing, this can only be fixed with a core change to expose this in some way via the REST API.

**Comment by rachelmcr** [comment]
Since this needs to be resolved on the backend, I'll close this issue and have reported this in https://github.com/woocommerce/woocommerce/issues/44085 for followup.

**Comment by rachelmcr** [comment]
Just a quick note that a [fix for this issue](https://github.com/woocommerce/woocommerce/pull/47784) has been merged, for release in WooCommerce core 9.1.0.

**Comment by Mehdi-YC** [comment]
Hello everyone, 
i just want to inform you that this is still not working when setting the quantuty to 0
the item is removed from the order, but the quantiy is not changed
